contracts/HeapOrdering.sol

```diff
@@ -0,0 +1,318 @@
+// SPDX-License-Identifier: GNU AGPLv3
+pragma solidity ^0.8.0;
+
+library HeapOrdering {
+    struct Account {
+        address id; // The address of the account.
+        uint256 value; // The value of the account.
+    }
+
+    struct HeapArray {
+        Account[] accounts; // All the accounts.
+        uint256 size; // The size of the heap portion of the structure, should be less than accounts length, the rest is an unordered array.
+        mapping(address => uint256) ranks; // A mapping from an address to a rank in accounts. Beware: ranks are shifted by one compared to indexes, so the first rank is 1 and not 0.
+    }
+
+    /// ERRORS ///
+
+    /// @notice Thrown when the address is zero at insertion.
+    error AddressIsZero();
+
+    /// INTERNAL ///
+
+    /// @notice Updates an account in the `_heap`.
+    /// @dev Only call this function when `_id` is in the `_heap` with value `_formerValue` or when `_id` is not in the `_heap` with `_formerValue` equal to 0.
+    /// @param _heap The heap to modify.
+    /// @param _id The address of the account to update.
+    /// @param _formerValue The former value of the account to update.
+    /// @param _newValue The new value of the account to update.
+    /// @param _maxSortedUsers The maximum size of the heap.
+    function update(
+        HeapArray storage _heap,
+        address _id,
+        uint256 _formerValue,
+        uint256 _newValue,
+        uint256 _maxSortedUsers
+    ) internal {
+        uint256 size = _heap.size;
+        uint256 newSize = computeSize(size, _maxSortedUsers);
+        if (size != newSize) _heap.size = newSize;
+
+        if (_formerValue != _newValue) {
+            if (_newValue == 0) remove(_heap, _id, _formerValue);
+            else if (_formerValue == 0) insert(_heap, _id, _newValue, _maxSortedUsers);
+            else if (_formerValue < _newValue) increase(_heap, _id, _newValue, _maxSortedUsers);
+            else decrease(_heap, _id, _newValue);
+        }
+    }
+
+    /// PRIVATE ///
+
+    /// @notice Computes a new suitable size from `_size` that is smaller than `_maxSortedUsers`.
+    /// @dev We use division by 2 to remove the leaves of the heap.
+    /// @param _size The old size of the heap.
+    /// @param _maxSortedUsers The maximum size of the heap.
+    /// @return The new size computed.
+    function computeSize(uint256 _size, uint256 _maxSortedUsers) private pure returns (uint256) {
+        while (_size >= _maxSortedUsers) _size >>= 1;
+        return _size;
+    }
+
+    /// @notice Returns the account of rank `_rank`.
+    /// @dev The first rank is 1 and the last one is length of the array.
+    /// @dev Only call this function with positive numbers.
+    /// @param _heap The heap to search in.
+    /// @param _rank The rank of the account.
+    /// @return The account of rank `_rank`.
+    function getAccount(HeapArray storage _heap, uint256 _rank)
```

Comment:
still don't find this function necessary. Moreover it differs from the other implementation which make comparison worse. `Heap` and `HeapOrdering` should be the most similar as possible without comprising on gas cost IMO

Reply:
I suggested this because it improves the readability of the code. Seeing account[rank-1] everywhere is kind of a nightmare for auditing purposes imo, and it doesn't add much gas cost.